drivers/sensor/paw3212/paw3212.c

```diff
@@ -0,0 +1,609 @@
+	if (err) {
+		return err;
+	}
+
+	k_delayed_work_init(&dev_data->init_work, paw3212_async_init);
+
+	k_delayed_work_submit(&dev_data->init_work,
+			      async_init_delay[dev_data->async_init_step]);
+
+	return err;
+}
+
+static int paw3212_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct paw3212_data *dev_data = &paw3212_data;
+	u8_t motion_status;
+	int err;
+
+	ARG_UNUSED(dev);
+
+	if (unlikely(!dev_data->ready)) {
+		LOG_INF("Device is not initialized yet");
+		return -EBUSY;
+	}
+
+	err = reg_read(dev_data, PAW3212_REG_MOTION, &motion_status);
+	if (err) {
+		LOG_ERR("Cannot read motion");
+		return err;
+	}
+
+	if ((motion_status & PAW3212_MOTION_STATUS_MOTION) != 0) {
+		u8_t x_low;
+		u8_t y_low;
+
+		if ((motion_status & PAW3212_MOTION_STATUS_DXOVF) != 0) {
+			LOG_WRN("X delta overlowed");
+		}
+		if ((motion_status & PAW3212_MOTION_STATUS_DYOVF) != 0) {
+			LOG_WRN("Y delta overlowed");
+		}
+
+		err = reg_read(dev_data, PAW3212_REG_DELTA_X_LOW, &x_low);
+		if (err) {
+			LOG_ERR("Cannot read X delta");
+			return err;
+		}
+
+		err = reg_read(dev_data, PAW3212_REG_DELTA_Y_LOW, &y_low);
+		if (err) {
+			LOG_ERR("Cannot read Y delta");
+			return err;
+		}
+
+		if (IS_ENABLED(CONFIG_PAW3212_12_BIT_MODE)) {
+			u8_t xy_high;
+
+			err = reg_read(dev_data, PAW3212_REG_DELTA_XY_HIGH,
+				       &xy_high);
+			if (err) {
+				LOG_ERR("Cannot read XY delta high");
+				return err;
+			}
+
+			dev_data->x = PAW3212_DELTA_X(xy_high, x_low);
+			dev_data->y = PAW3212_DELTA_Y(xy_high, y_low);
+		} else {
+			dev_data->x = (s8_t)x_low;
+			dev_data->y = (s8_t)y_low;
+		}
+	} else {
+		dev_data->x = 0;
+		dev_data->y = 0;
+	}
+
+	return err;
+}
+
+static int paw3212_channel_get(struct device *dev, enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct paw3212_data *dev_data = &paw3212_data;
+
+	ARG_UNUSED(dev);
+
+	if (unlikely(!dev_data->ready)) {
+		LOG_INF("Device is not initialized yet");
+		return -EBUSY;
+	}
+
+	switch (chan) {
+	case SENSOR_CHAN_POS_DX:
+		val->val1 = dev_data->x;
+		val->val2 = 0;
+		break;
+
+	case SENSOR_CHAN_POS_DY:
+		val->val1 = dev_data->y;
+		val->val2 = 0;
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int paw3212_trigger_set(struct device *dev,
+			       const struct sensor_trigger *trig,
+			       sensor_trigger_handler_t handler)
+{
+	struct paw3212_data *dev_data = &paw3212_data;
+	int err;
+
+	ARG_UNUSED(dev);
+
+	if (unlikely(trig->type != SENSOR_TRIG_DATA_READY)) {
+		return -ENOTSUP;
+	}
+
+	if (unlikely(trig->chan != SENSOR_CHAN_ALL)) {
+		return -ENOTSUP;
+	}
+
+	if (unlikely(!dev_data->ready)) {
+		LOG_INF("Device is not initialized yet");
+		return -EBUSY;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&dev_data->lock);
+
+	if (handler) {
+		err = gpio_pin_enable_callback(dev_data->irq_gpio_dev,
+					       CONFIG_PAW3212_IRQ_GPIO_PIN);
+	} else {
+		err = gpio_pin_disable_callback(dev_data->irq_gpio_dev,
```

Comment:
Wouldn't the following:
```
	err = gpio_pin_disable_callback(dev_data->irq_gpio_dev,
				       CONFIG_PAW3212_IRQ_GPIO_PIN);
	if (!err && handler) {
		dev_data->data_ready_handler = handler;
		err = gpio_pin_enable_callback(dev_data->irq_gpio_dev,
					       CONFIG_PAW3212_IRQ_GPIO_PIN);
	}
```
be easier and a bit safer? I seems that right now the interrupt may become pending while its execution is locked, and in case the handler is deactivated, after the interrupts are unlocked, the callback will try to call `NULL`.

Reply:
On second thought, I think that I went too far, `NULL` cannot be called, as the execution of the callback should be filtered out (since it was disabled) in the GPIO interrupt, even if the interrupt was executed. But still, I think there is no need to lock interrupts here.

Reply:
I don't think that `gpio_pin_enable_callback` or `gpio_pin_disable_callback` work as an implicit barrier. If these are not then the only safe way to have `data_ready_handler` set it under a lock with the `gpio_pin_enable_callback`. Note that compiler can reorder `dev_data->data_ready_handler = handler;` as it wishes.

I prefer to leave it as is.

Reply:
You can prevent the reordering by marking `data_ready_handler` as `volatile`.
Anyway, if you prefer to keep the current implementation, I won't insist on the change, it was a suggestion only. The code is not incorrect, at most a bit too "heavy".

Reply:
> You can prevent the reordering by marking data_ready_handler as volatile.

Actually this will not work as compiler will only keep order related to other volatile accesses. If you mark only one variable as volatile it can be reordered in relation to anything non-volatile. The only thing that will protest against it as a compiler barrier.